firedrake/variational_solver.py

```diff
@@ -281,7 +289,12 @@ def __init__(self, a, L, u, bcs=None, aP=None,
             F = ufl_expr.action(J, u) - L
 
         super(LinearVariationalProblem, self).__init__(F, u, bcs, J, aP,
-                                                       form_compiler_parameters=form_compiler_parameters)
+                                                       form_compiler_parameters=form_compiler_parameters,
+                                                       is_linear=True)
+        # check all equations are linear
+        for bc in self.bcs:
+            assert bc.is_linear, "All equations including ones defined in EquationBC objects have to be linear"
```

Comment:
Would `assert isinstance(bc, LinearVariationalProblem)` be sufficient?

Reply:
`bc` is currently not an instance of `LinearVariationalProblem`, but an instance of `EquationBC.`  I have an idea of extracting common components of `EquationBCSplit` and `NonlinearVariationalProblem` (such as `F`, `J`, `Jp`, and `bcs`), which will also simplify this part.  I would like to discuss this idea some time, but as it requires many structural change, I'm not going to do it for this PR.